### DISCUSSION TITLE
TCP Client Issue in Setup

### DISCUSSION THREAD

**Discussion by mosa11aei** [nasa/fprime#2134] [Q&A] [answered]
I am trying to set up a TCP client through F', and I'm modeling it off of the setup for the TCP client setup for the default communication driver in-built. I am able to connect to the TCP server and receive data, but I'm getting a weird error when trying to send data. The following assertion is failing:

Svc > StaticMemory > bufferDeallocate_handler
```cpp
FW_ASSERT((fwBuffer.getData() + fwBuffer.getSize()) <= (m_static_memory[portNum] + sizeof(m_static_memory[0])), fwBuffer.getSize(), sizeof(m_static_memory[0]));
```
I can't really make heads or tails of it, but I think it has to do with memory runoff. As in, it's trying to deallocate memory that wasn't allocated in the first place, which is a bit confusing. Here is how I am trying to send data to the TCP driver I've made, named `py2c`:
```cpp
      Fw::String testString("Hello");

      Fw::Buffer sendBuffer;
      sendBuffer.setData(reinterpret_cast<U8 *>(const_cast<char *>(testString.toChar())));
      sendBuffer.setSize(testString.length());

      Drv::SendStatus status = this->dataOut_out(0, sendBuffer);

      if (status.e != Drv::SendStatus::SEND_OK)
      {
        Fw::Logger::logMsg("[BME680] Failed to send packet to TCP.\n");
      }
      else
      {
        Fw::Logger::logMsg("[BME680] One packet sent to TCP.\n");
        this->dataInDeallocate_out(0, sendBuffer);
      } 
```

Here's my `topology.fpp` file:
```cpp
      py2c.allocate -> staticMemory.bufferAllocate[Ports_StaticMemory.py2c] // defined in the StaticMemory enum
      py2c.$recv -> bme680.dataIn
      
      bme680.dataOut -> py2c.send      
      bme680.dataInDeallocate -> staticMemory.bufferDeallocate[Ports_StaticMemory.py2c]
      py2c.deallocate -> staticMemory.bufferDeallocate[Ports_StaticMemory.py2c]
```

...and my Topology.cpp file:
```cpp
    try {
        Os::TaskString name("Py2c");
        Fw::String py2chost("127.0.0.1");
        py2c.configure(py2chost.toChar(), 7860);
        py2c.startSocketTask(name, true, COMM_PRIORITY, Default::STACK_SIZE);
        
        logger.log("[INFO] Connected to Py2c TCP server on port 7860.\n");
    } catch(...) {
        logger.log("[ERROR] Failed to start py2c socket task\n");
    }
```

**Comment by ephraim271** [comment]
@mosa11aei 
Did you find a fix for this? I'm trying something similar

**Reply by mosa11aei** [reply]
What specifically? Sending from F Prime over TCP, or getting a TCP link?

**Comment by zimri-leisher** [comment, marked as answer]
Funnily enough, this is a very similar issue to another issue that was just posted. In general, this problem is caused by passing a buffer to StaticMemory's bufferDeallocate port that didn't come from StaticMemory's bufferAllocate port. You can think of Buffers as being required to follow a loop, starting from StaticMemory (or my preferred alternative, BufferManager), going through various components, and ending up at StaticMemory/BufferManager again.

I would take a look at the setup and my answer for #3079 .

In general, the flow you want is:
* Inside component X that wants to send something:
  * Get a buffer from a buffer manager/static memory
  * Serialize your data into the buffer
  * Send the buffer to a driver
* The driver should be connected to the same buffer manager/static memory that your component X is connected to
  * The driver, on sending the data via a socket, will automatically call deallocate for you.
  * IMPORTANT CAVEAT: for the TcpClient, if the driver returns Drv::SendStatus::SEND_RETRY, then it will not deallocate the buffer, so you have to check for this and do it yourself in this case.

**Reply by ephraim271** [reply]
@zimri-leisher Many thanks, this worked!
Is this behaviour documented somewhere/ which document should I be referring to? 
I was reading https://github.com/nasa/fprime/blob/devel/Drv/TcpClient/docs/sdd.md and it just mentioned Fw::Buffer and not Buffer Manager :sweat_smile:

**Reply by zimri-leisher** [reply]
Awesome! If that fixed it, can you mark it as the correct answer? Thanks. EDIT--never mind, I see you aren't the original author.

Hmm, the only obvious place I know it's shown is in the Ref topology.
